yjit/src/codegen.rs

```diff
@@ -2513,14 +2504,18 @@ fn gen_equality_specialized(
         }
 
         // Call rb_str_eql_internal(a, b)
-        call_ptr(cb, REG0, rb_str_eql_internal as *const u8);
+        let val = asm.ccall(rb_str_eql_internal as *const u8, vec![a_opnd, b_opnd]);
 
         // Push the output on the stack
-        cb.write_label(ret);
         ctx.stack_pop(2);
         let dst = ctx.stack_push(Type::UnknownImm);
-        mov(cb, dst, RAX);
-        cb.link_labels();
+        asm.mov(dst, val);
+        asm.jmp(ret);
+
+        asm.write_label(equal);
+        asm.mov(dst, Opnd::Imm(Qtrue.into()));
```

Comment:
In general we really try to avoid having branches inside the generated code because it's potentially bad for performance (could stall the pipeline). It's best that the code be as flat/linear as possible.

I think it's good to push to the stack in two separate places, it makes for more readable code IMO.

Reply:
I tried using `C_RET_OPND` and it worked https://github.com/Shopify/ruby/pull/371/commits/99381765d0008ff0f03ea97c6c8db608a2298e2b, but I reverted it as it seems hacky as Maxime said https://github.com/Shopify/ruby/pull/371#discussion_r939038932.